### PR TITLE
[JENKINS-22896] Allow specifying a different job for test results.

### PR DESCRIPTION
**Pull request by abayer**
https://issues.jenkins-ci.org/browse/JENKINS-22896

This allows you to specify a different job for the test results used
to split the parallel test execution. With this functionality, you can
use this with multibranch jobs (say, pull requests) pointing at a job
with a longer lifespan than the more ephemeral PR jobs, so that the
first build of a PR job doesn't need to run all the tests due to lack
of history.

cc @reviewbybees, especially @jglick

### CODE REVIEW

File: src/main/java/org/jenkinsci/plugins/parallel_test_executor/ParallelTestExecutor.java [hunk 1]
```diff
@@ -274,9 +293,28 @@ static private void collect(TestResult r, Map<String, TestClass> data) {
...
-    private static TestResult findPreviousTestResult(Run<?, ?> b, TaskListener listener) {
-        for (int i = 0; i < NUMBER_OF_BUILDS_TO_SEARCH; i++) {// limit the search to a small number to avoid loading too much
-            b = b.getPreviousBuild();
+    private static TestResult findPreviousTestResult(Run<?, ?> originalBuild, TaskListener listener, String alternateJob) {
+        Run<?,?> b = originalBuild.getPreviousBuild();
+        if (alternateJob != null && !alternateJob.equals("")) {
+            TopLevelItem item = Jenkins.getInstance().getItem(alternateJob);
+            if (item != null) {
+                if (item instanceof Job) {
+                    Job j = (Job) item;
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/parallel_test_executor/ParallelTestExecutor.java]: :bug: no.

```java
Job j = Jenkins.getInstance().getItem(alternateJob, originalBuild.getParent(), Job.class);
```

File: src/main/java/org/jenkinsci/plugins/parallel_test_executor/ParallelTestExecutor.java [hunk 2]
```diff
@@ -303,6 +342,10 @@ public AutoCompletionCandidates doAutoCompleteTestJob(@QueryParameter String val
             return AutoCompletionCandidates.ofJobNames(AbstractProject.class, value, self, container);
         }
 
+        public AutoCompletionCandidates doAutoCompleteAlternateJob(@QueryParameter String value, @AncestorInPath Item self, @AncestorInPath ItemGroup container) {
+            return AutoCompletionCandidates.ofJobNames(AbstractProject.class, value, self, container);
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/parallel_test_executor/ParallelTestExecutor.java]: :bug: `Job`

File: src/main/java/org/jenkinsci/plugins/parallel_test_executor/SplitStep.java [hunk 3]
```diff
@@ -34,6 +36,15 @@ public Parallelism getParallelism() {
         return parallelism;
     }
 
+    public String getAlternateJob() {
+        return alternateJob;
+    }
+
+    @DataBoundSetter
+    public void setAlternateJob(String alternateJob) {
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/parallel_test_executor/SplitStep.java]: You forgot `doAutoCompleteAlternateJob`.

File: src/main/java/org/jenkinsci/plugins/parallel_test_executor/ParallelTestExecutor.java (hunk 1 above)
Comment (abayer) [inline src/main/java/org/jenkinsci/plugins/parallel_test_executor/ParallelTestExecutor.java]: Okiedokie!

File: src/main/java/org/jenkinsci/plugins/parallel_test_executor/SplitStep.java (hunk 3 above)
Comment (abayer) [inline src/main/java/org/jenkinsci/plugins/parallel_test_executor/SplitStep.java]: ...whoops.

**Comment by jglick** [comment]
I am unclear on the motivation here.

The JIRA issue seems to be referring to freestyle builds. Indeed the hackish system needed for freestyle is not very reliable. But the title seems to be misleading. They want to choose which _build_ to use as the reference, not which _job_. Possibly the reporter is unaware of the distinction in terminology.

Then your stated motivation here is about multibranch Pipeline projects. But there is no test for that. And anyway this would be the wrong solution. See [JENKINS-33274](https://issues.jenkins-ci.org/browse/JENKINS-33274).

**Comment by abayer** [comment]
Ok, then I'll open another JIRA for this, and will add a Pipeline test. But I think it's still worth doing unless [JENKINS-33274](https://issues.jenkins-ci.org/browse/JENKINS-33274) has a fix coming soon, since this plugin is currently useless for multibranch PR builds.

**Comment by jglick** [comment]
Well then let us prioritize work on JENKINS-33274, rather than coming up with hackish workarounds which we cannot later delete due to settings compatibility issues.

**Comment by abayer** [comment]
Alright, I'll take a look at JENKINS-33274.

**Comment by Kukunin** [comment]
Hello. First of all, thanks for the great plugin.

I've spent 3 full days to migrate our Jenkins to Pipeline (Multibranch job) with support of parallel-test-executor plugin. 

But this is very disappointed fact, that this plugin still can't use reports from neighbour jobs to parallel jobs. 

Just want to express 👍 here

**Comment by sford** [comment]
[JENKINS-33274](https://issues.jenkins-ci.org/browse/JENKINS-33274) looks promising to solve this 👍 

We are currently working around this by moving our large tests that we want to parallelize to their own standalone pipeline job. This job is then used by a multibranch project via a `build` step. This allows us to take advantage of multibranch as much as we can while allowing us to parallelize a subset of our tests in a separate job that maintains history.

Not optimal having pipeline logic split, but the benefits of parallelizing our tests are worth it.

**Comment by roni-frantchi** [comment]
@abayer thanks so much for this.
@tolik-u this is exactly what we need!

**Comment by abayer** [comment]
Closing [JENKINS-33274](https://issues.jenkins-ci.org/browse/JENKINS-33274) is the right direction to go in the long run.

**Pull request closed without merging**